File: app/src/main/java/com/refrii/client/FoodListAdapter.java
package com.refrii.client;

import android.content.Context;
import android.support.design.widget.Snackbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.daimajia.swipe.SwipeLayout;
import com.daimajia.swipe.adapters.BaseSwipeAdapter;

import java.text.SimpleDateFormat;
import java.util.Collections;
import java.util.List;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by yusuke on 2017/09/17.
 */

public class FoodListAdapter extends BaseSwipeAdapter {

    private Context mContext;
    private List<Food> mFoods;
    private FoodListAdapter mFoodListAdapter;

    public FoodListAdapter(Context context, List<Food> foods) {
        mContext = context;
        mFoods = foods;
        mFoodListAdapter = this;
        Collections.sort(foods);
    }

    public void add(Food food) {
        this.mFoods.add(food);
    }

    public void remove(Food food) {
        mFoods.remove(food);
    }

    @Override
    public int getSwipeLayoutResourceId(int position) {
        return R.id.foodListSwipeLayout;
    }

    @Override
    public View generateView(final int position, ViewGroup parent) {
        View v = LayoutInflater.from(mContext).inflate(R.layout.food_list_row, null);
        final SwipeLayout swipeLayout = (SwipeLayout)v.findViewById(getSwipeLayoutResourceId(position));
        swipeLayout.setOnDoubleClickListener(new SwipeLayout.DoubleClickListener() {
            @Override
            public void onDoubleClick(SwipeLayout layout, boolean surface) {
                Toast.makeText(mContext, "DoubleClick", Toast.LENGTH_SHORT).show();
            }
        });

        v.findViewById(R.id.incrementImageView).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Food food = mFoods.get(position);
                food.setAmount(food.getAmount() + 1);
                updateFood(food, view);
            }
        });
        v.findViewById(R.id.decrementImageView).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Food food = mFoods.get(position);
                food.setAmount(food.getAmount() - 1);
                updateFood(food, view);
            }
        });
        return v;
    }

    @Override
    public void fillValues(int position, View convertView) {
        Food food = mFoods.get(position);
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy/MM/dd");

        ((TextView)convertView.findViewById(R.id.nameFoodListTextView)).setText(food.getName());
        ((TextView)convertView.findViewById(R.id.expirationDateFoodListTextView)).setText(formatter.format(food.getExpirationDate()));
        ((TextView)convertView.findViewById(R.id.amountFoodListTextView)).setText(String.valueOf(food.getAmount() + " " + food.getUnit().getLabel()));
    }

    @Override
    public int getCount() {
        return this.mFoods.size();
    }

    @Override
    public Object getItem(int position) {
        return this.mFoods.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    private void updateFood(Food food, final View view) {
        RequestBody body = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("amount", String.valueOf(food.getAmount()))
                .build();
        FoodService service = RetrofitFactory.getClient(FoodService.class, mContext);
        Call<Food> call = service.updateFood(food.getId(), body);
        call.enqueue(new Callback<Food>() {
            @Override
            public void onResponse(Call<Food> call, Response<Food> response) {
                Food food = response.body();
                Snackbar.make(view, "Amount of " + food.getName() + " updated", Snackbar.LENGTH_LONG)
                        .setAction("Dismiss", null)
                        .show();
                mFoodListAdapter.notifyDataSetChanged();
            }

            @Override
            public void onFailure(Call<Food> call, Throwable t) {
                Snackbar.make(view, t.getMessage(), Snackbar.LENGTH_LONG)
                        .setAction("Dismiss", null)
                        .show();
            }
        });
    }
}

File: app/src/main/java/com/refrii/client/views/adapters/FoodListAdapter.kt
package com.refrii.client.views.adapters

import android.content.Context
import android.support.design.widget.Snackbar
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import com.daimajia.swipe.SwipeLayout
import com.daimajia.swipe.adapters.BaseSwipeAdapter
import com.refrii.client.R
import com.refrii.client.models.Food
import com.refrii.client.services.FoodService
import com.refrii.client.services.RetrofitFactory
import io.realm.Realm
import io.realm.RealmConfiguration
import okhttp3.MultipartBody
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.text.SimpleDateFormat
import java.util.*

class FoodListAdapter(
        private val mContext: Context,
        private val mFoods: MutableList<Food>) : BaseSwipeAdapter() {

    private val mFoodListAdapter: FoodListAdapter = this
    private var mRealm: Realm

    init {
        Collections.sort(mFoods)

        Realm.setDefaultConfiguration(RealmConfiguration.Builder(mContext).build())
        mRealm = Realm.getDefaultInstance()
    }

    fun add(food: Food) {
        this.mFoods.add(food)
    }

    fun remove(food: Food) {
        mFoods.remove(food)
    }

    override fun getSwipeLayoutResourceId(position: Int): Int = R.id.foodListSwipeLayout

    override fun generateView(position: Int, parent: ViewGroup): View {
        val v = LayoutInflater.from(mContext).inflate(R.layout.food_list_row, null)
        val swipeLayout = v.findViewById<View>(getSwipeLayoutResourceId(position)) as SwipeLayout

        swipeLayout.setOnDoubleClickListener { _, _ -> Toast.makeText(mContext, "DoubleClick", Toast.LENGTH_SHORT).show() }

        v.findViewById<View>(R.id.incrementImageView).setOnClickListener { view ->
            val food = mFoods[position]

            food.amount = food.amount + 1
            updateFood(food, view)
            syncFood(food)
        }

        v.findViewById<View>(R.id.decrementImageView).setOnClickListener { view ->
            val food = mFoods[position]
            food.amount = food.amount - 1
            updateFood(food, view)
            syncFood(food)
        }

        return v
    }

    override fun fillValues(position: Int, convertView: View) {
        val food = mFoods[position]
        val formatter = SimpleDateFormat("yyyy/MM/dd", Locale.getDefault())
        val nameFoodListTextView = convertView.findViewById<View>(R.id.nameFoodListTextView) as TextView
        val expirationDateFoodListTextView = convertView.findViewById<View>(R.id.expirationDateFoodListTextView) as TextView
        val amountFoodListTextView = convertView.findViewById<View>(R.id.amountFoodListTextView) as TextView

        nameFoodListTextView.text = food.name
        expirationDateFoodListTextView.text = formatter.format(food.expirationDate)
        amountFoodListTextView.text = food.amount.toString() + " " + food.unit!!.label
    }

    override fun getCount(): Int = this.mFoods.size

    override fun getItem(position: Int): Any = this.mFoods[position]

    override fun getItemId(position: Int): Long = position.toLong()

    private fun updateFood(food: Food, view: View) {
        mRealm.executeTransaction { mRealm.copyToRealmOrUpdate(food) }
        mFoodListAdapter.notifyDataSetChanged()

        Snackbar.make(view, "Updated successfully.", Snackbar.LENGTH_LONG).show()
    }

    private fun syncFood(food: Food) {
        val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("amount", food.amount.toString())
                .build()

        RetrofitFactory.getClient(FoodService::class.java, mContext)
                .updateFood(food.id, body)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object: Subscriber<Food>() {
                    override fun onError(e: Throwable) {
                        Toast.makeText(mContext, e.message, Toast.LENGTH_LONG).show()
                    }

                    override fun onCompleted() {
                        Log.d(TAG, "Update completed.")
                    }

                    override fun onNext(t: Food) {
                        mFoodListAdapter.notifyDataSetChanged()
                    }
                })
    }

    companion object {
        private const val TAG = "FoodListAdapter"
    }
}
